File: python/findprimes.py
import math
import argparse
import signal
import sys

# Global variables to allow signal_handler to reach them.
primes = []
num_cycles = 0


def findprimes(limit) -> None:
    """
    Looks for prime numbers up to the limit.
    """
    global primes
    global num_cycles

    primes = [2]  # Reset the list of primes
    # Check all odd numbers as even numbers can't be prime
    for i in range(3, int(limit), 2):
        is_prime = True
        sqrt_i = math.sqrt(i)
        for _, j in enumerate(primes):  # Iterate over all already known primes
            # Exit if prime is larger than sqrt of i as it cannot divide evenly any longer.
            if j > sqrt_i:
                break

            # If j divides i, i isn't prime.
            if i % j == 0:
                is_prime = False
                break

        # Add new prime to list of primes
        if is_prime:
            primes.append(i)

    # Increment number of fully completed cycles
    num_cycles += 1


def signal_handler(sig, frame) -> None:
    """
    Handles SIGINT/ CTRL+C.
    """
    global primes
    global num_cycles

    # Print "number_cycles - current_length_of_primes"
    print(f"{num_cycles} - {len(primes)}", end="")
    sys.exit(0)


def main():
    """
    main()
    """

    # Get the upper limit from command line args
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "integers", metavar="N", type=int, nargs="+", help="an upper limit to go to"
    )
    args = parser.parse_args()

    # Set CTRL+C/Sigint handler
    signal.signal(signal.SIGINT, signal_handler)

    # Keep running until program is manually interrupted
    while True:
        findprimes(args.integers[0])


if __name__ == "__main__":
    main()

File: java/FindPrimes.java
import java.util.ArrayList;
import java.lang.Math;

public class FindPrimes
{
    public static void main(String[] args) {
		int limit = Integer.parseInt(args[0]);
		final long startTime = System.currentTimeMillis();

		calculatePrimes(limit);

		final long endTime = System.currentTimeMillis();
		System.out.println("Java execution time: " + (endTime - startTime) + "ms");
    }

    public static void calculatePrimes(int limit) {
	// Create a vector to store known primes
		ArrayList<Integer> primes = new ArrayList<Integer>();
		primes.add(2); // Add first known prime

		for (int i=3; i < limit; i++) { // Iterate over all numbers from 2 to 10,000,000
			boolean isPrime = true;
			double sqrt_i = Math.sqrt(i);
			
			for (int j = 0; j < primes.size(); j++) {
				if (primes.get(j) > sqrt_i) { // Max value to compare with (due to how math works)
					break;
				}
					
				if (i % primes.get(j) == 0) { // If its divisible by any already known prime
					isPrime = false;
					break;
				}
			}
			if (isPrime == true) {
				primes.add(i); // Add to known primes
			}
		}
    }
}
